controller/sync.go

```diff
@@ -346,7 +347,8 @@ func CreateTwoWayMergePatch(orig, new, dataStruct interface{}) ([]byte, bool, er
 }
 
 // persistRolloutStatus persists updates to rollout status. If no changes were made, it is a no-op
-func (c *Controller) persistRolloutStatus(orig *v1alpha1.Rollout, newStatus *v1alpha1.RolloutStatus, newPause *bool) error {
+func (c *Controller) persistRolloutStatus(orig *v1alpha1.Rollout, newStatus *v1alpha1.RolloutStatus, newPause *bool, phase metrics.ReconcilePhase) error {
+	c.metricsServer.IncPhase(orig, phase)
```

Comment:
I think `RolloutStatus` has or supposed to have all information which describes current state of rollout. Instead of imperatively set it in code we can infer the state within `persistRolloutStatus` method. Most likely you can check `InvalidSpec` and `Available` condition and make decision about current phase. What do you think?

Reply:
I like this approach a lot more.  The only issue is that the `Available` condition needs to be cleaned up (as per #41), but I think it makes sense to go start with this implementation and work on that #41 afterward.

Reply:
Sure